### ISSUE TITLE
Dismissing keyboard programatically

### ISSUE THREAD

**Issue by mikemimik** [flutter/flutter#7247]
It's not clear how to accomplish this.

**Comment by eseidelGoogle** [comment]
@Hixie or @mpcomplete presumably this is a docs bug?

**Comment by mikemimik** [comment]
When you select a `InputFormField` the keyboard automatically [gets requested](https://github.com/flutter/flutter/blob/7d346dd1840adf8f9ef9bfcddeca2ac908a91da7/packages/flutter/lib/src/material/input.dart#L120) and is shown. There is actually no way to make it dismiss from the UI and it's unclear or not possible to do it from current widgets.

**Comment by Hixie** [comment]
From the UI it's easy, just press the Android OS-level down arrow button (replaces the back button when the keyboard is up).

**Comment by mikemimik** [comment]
I'm working in iOS and there is no such OS-level trigger. I tested in the simulator and on device, I get the same standard keyboard.

**Comment by chinmaygarde** [comment]
[`TextInputConnection.close`](https://docs.flutter.io/flutter/services/TextInputConnection/close.html) should do what you need.

**Comment by mikemimik** [comment]
@chinmaygarde The `.close()` method is not static on that class, which means I need to access it on an instance member. However the constructor is private, so I can't create an instance member.

**Comment by mikemimik** [comment]
I figured out how to do it.

Inside a section of code which has access to the `context` you just need to have `Focus.clear(context);` and it will dismiss the keyboard.

**Comment by mikemimik** [comment]
I'm come across an error with the about implementation. As you'll see from the below images, I'm using code to dismiss the keyboard by calling `Focus.clear(context);`.
> I'm not sure how unfocusing dismisses the keyboard, as in, precisely what flutter is doing under the hood.

The effect of being 'focused' is still in effect on the `InputFormField` widget.

> The code to achieve these images is in [this repo](https://github.com/mikemimik/flutter-examples/tree/master/form_example).

### Unselected input:
![simulator screen shot dec 15 2016 3 21 07 am](https://cloud.githubusercontent.com/assets/2818462/21216614/c3a928fe-c275-11e6-9ffc-5ebd7559ad2f.png)

### Selected input:
![simulator screen shot dec 15 2016 3 21 19 am](https://cloud.githubusercontent.com/assets/2818462/21216636/dc8d9080-c275-11e6-8d9f-ca8e786181ae.png)

### Dismissed keyboard:
![simulator screen shot dec 15 2016 3 22 12 am](https://cloud.githubusercontent.com/assets/2818462/21216644/e768c38a-c275-11e6-9fc5-9e1900c72644.png)

**Comment by sethladd** [comment]
@mikemimik did you ever get this to work? What was your recipe?

**Comment by mikemimik** [comment]
@sethladd Will cycle back to this on the weekend. Apologies for being so terribly behind on responding.

**Comment by pedia** [comment]
I found HACK way:

> import 'package:flutter/services.dart';

and call

> SystemChannels.textInput.invokeMethod('TextInput.hide');

**Comment by jonasbark** [comment]
This is a better way, as it also clears the focus:
`FocusScope.of(context).requestFocus(new FocusNode());`
from
https://stackoverflow.com/a/44991969/1068346

**Comment by kentcb** [comment]
Another problem here (even with the hacks) is knowing _when_ the keyboard has been dismissed. The `requestFocus` method does not return a `Future`, so there's no easy way to wait for the dismissal.

This is a problem for me because I want the keyboard to dismiss so I can accurately determine where to show a menu. I use `RenderBox` to figure this out, but the keyboard remaining on screen screws up measurements. If I stick an `await new Future.delayed(...)` after dismissing the keyboard and before the measurement code, it works as long as the delay is long enough (who knows what long enough really is though...)

**Comment by Sun3** [comment]
What we really need especially on the iOS side is a way to add a Done button to the keyboard.

**Comment by eseidelGoogle** [comment]
@Sun3 would you be willing to file a bug for a Done button with an example from an iOS app you're seeking to replicate?

**Comment by zoechi** [comment]
@sun3 @eseidel might be covered by #9573

**Comment by sahil311289** [comment]
Did anyone try this:
FocusScope.of(context).requestFocus(new FocusNode());
I read it from another forum.

**Comment by manujbahl** [comment]
It has stopped working for me now.

**Comment by manujbahl** [comment]
As requested here is the code.

```
main() {
    runApp(new Home());
}


class Home extends StatelessWidget {
    @override
      Widget build(BuildContext context) {
        var widget = new MaterialApp(
            home: new Scaffold(
                body: new Container(
                    height:500.0,
                    child: new GestureDetector(
                        onTap: () {
                            FocusScope.of(context).requestFocus(new FocusNode());
                        },
                        child: new Container(
                            color: Colors.white,
                            child:  new Column(
                                mainAxisAlignment:  MainAxisAlignment.center,
                                crossAxisAlignment: CrossAxisAlignment.center,
                                
                                children: [
                                    new TextField( ),
                                    new Text("Test"),                                
                                ],
                            )
                        )
                    )
                )
            ),
        );
           
        return widget;
    }
}
```


Here is the flutter doctor result .
Manujs-MacBook-Pro:Examples MBahl$ flutter doctor -v
[✓] Flutter (Channel master, v0.5.5-pre.14, on Mac OS X 10.13.5 17F77, locale en-US)
    • Flutter version 0.5.5-pre.14 at /Users/MBahl/flutter/flutter
    • Framework revision b401e76554 (5 weeks ago), 2018-06-13 09:20:19 -0700
    • Engine revision c3976b3c71
    • Dart version 2.0.0-dev.61.0.flutter-c95617b19c

[✓] Android toolchain - develop for Android devices (Android SDK 28.0.1)
    • Android SDK at /Users/MBahl/Library/Android/sdk
    • Android NDK at /Users/MBahl/Library/Android/sdk/ndk-bundle
    • Platform android-28, build-tools 28.0.1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 9.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.4.1, Build version 9F2000
    • ios-deploy 1.9.2
    • CocoaPods version 1.5.3

[✓] Android Studio (version 3.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 26.0.1
    • Dart plugin version 173.4700
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)

[!] VS Code (version 1.25.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected devices (1 available)
    • Manuj's iPhone • 4b185da29e7eb1d7e37d22ce0a205e7c081ca5b8 • ios • iOS 11.3.1

! Doctor found issues in 1 category.



YES  "stopped working" means ---  doesn't release focus and the keyboard stays on screen?

**Comment by jonsamwell** [comment]
This is working on the latest dev build (v0.5.7).

**Comment by manujbahl** [comment]
just upgraded and still does not work .. 
Flutter (Channel master, v0.5.8-pre.79, on Mac OS X 10.13.5 17F77, locale en-US)

**Comment by Shantanu-Kotambkar** [comment]
- @manujbahl `FocusScope.of(context).requestFocus(new FocusNode()); `, worked for me. Thanks for the solution. 

- Here's a tip @manujbahl you are on master channel, that's probably the problem you are facing! You should change your channel to Beta and upgrade your flutter installation. Tell us if you run in any trouble. Happy to help!

**Comment by manujbahl** [comment]
@Shantanu-Kotambkar Did you try the above code on beta channel? What are the other differences between master and beta? I just want to make sure nothing else breaks if move to master.

**Comment by Shantanu-Kotambkar** [comment]
Yes I did try, I feel it is better to use beta channel as it has many updates that remove some specific bugs. I am developing using Beta channel and so far nothing is breaking on my end. I would recommend switching to Beta.

**Comment by manujbahl** [comment]
just did and still does not work. What version are you using.. and I am trying out on ios

**Comment by Shantanu-Kotambkar** [comment]
Here you go : 

> $ flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel beta, v0.5.1, on Mac OS X 10.13.6 17G65, locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK 28.0.0)
[✓] iOS toolchain - develop for iOS devices (Xcode 9.0.1)
[✓] Android Studio (version 3.1)
[!] VS Code (version 1.24.0)
[✓] Connected devices (1 available)

! Doctor found issues in 1 category.

**Comment by manujbahl** [comment]
Why should VS Code version matter ?

**Comment by Shantanu-Kotambkar** [comment]
No it doesn't matter
Show us your flutter doctor result @manujbahl

**Comment by manujbahl** [comment]
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel beta, v0.5.1, on Mac OS X 10.13.5 17F77, locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK 28.0.1)
[✓] iOS toolchain - develop for iOS devices (Xcode 9.4.1)
[✓] Android Studio (version 3.1)
[!] VS Code (version 1.25.1)
[✓] Connected devices (1 available)

**Comment by Shantanu-Kotambkar** [comment]
`       onTap: () {
                            FocusScope.of(context).requestFocus(new FocusNode());
                        },`, I think you are implementing it wrong here. 

Where do you tap to trigger this onTap call

**Comment by manujbahl** [comment]
any luck ?

**Comment by manujbahl** [comment]
Is anyone planning on fixing this ? I just upgraded to the latest beta channel but no luck

**Comment by eoinfogarty** [comment]
`FocusScope.of(context).requestFocus(new FocusNode())` works.

Just a note for anybody using a `SingleChildScrollView` make sure that `GestureDetector` wraps the the scroll view and not the other way round.

``` dart
class _HomePageState extends State<HomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: GestureDetector(
        onTap: () {
          FocusScope.of(context).requestFocus(FocusNode());
        },
        child: Scrollbar(
          child: SingleChildScrollView(
            child: Container(
              width: double.maxFinite,
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  TextField(),
                  TextField(),
                  TextField(),
                  TextField(),
                  TextField(),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}
```

Putting `GestureDetector` inside the scrollview seems to prevent `GestureDetector` catching some taps.

**Comment by AppGodOrg** [comment]
I don't know if this will help anyone but adding this line to GestureDetector solved my issues capturing the tap outside of the text field:

`behavior: HitTestBehavior.opaque,`

**Comment by SteveAlexander** [comment]
@manujbahl you're getting `FocusScope.of(context)` where the context is taken from below your scaffold and material components. Split this into separate components, or use a `Builder` and you'll find that this works.

**Comment by moadqassem** [comment]
@manujbahl The problem with the keyboard doesn't work for everyone, for example losing the focus from a text field and assign it to unused focus node doesn't really work(at least for me):
   
` FocusScope.of(context).requestFocus(new FocusNode());`

One of the solution that could solve your problem is using the SystemChannel where you act on the text input channel and hide it once the Input Field loses the focus: 

```
            body: GestureDetector(
                onTap: () {
                  SystemChannels.textInput.invokeMethod('TextInput.hide');
                },

```

Hope will solve your problem.

**Comment by zoechi** [comment]
@moadqassem that is known to have stopped working with API 28. If you switch back to 27 it should work again as a workaround until this issue is fixed.

**Comment by psine** [comment]
This issue has been open for a long time and I'm not sure how the framework might have changed in the interim, but I can dismiss the keyboard for a TextField (formerly known as InputFormField) widget instance, "textField", by first instantiating a named FocusNode, "focusNode," passing it to the textField constructor, then handling `textField.onSubmitted` by calling `focusNode.unfocus()`.  I think this is cleaner than trying to set the focus to a new, anonymous FocusNode.

**Comment by Frioo** [comment]
@jonasbark Thank you for the solution, it works fine for me (although I feel like a bastard every time I use such workarounds)

for anyone else: `FocusScope.of(context).requestFocus(new FocusNode());` is the way to go

**Comment by diegoveloper** [comment]
I created a package that you can use.

![keyboard_custom_actions](https://user-images.githubusercontent.com/4898256/49628648-57111680-f9b3-11e8-937d-1beeb1bb677f.gif)

https://pub.dartlang.org/packages/keyboard_actions

**Comment by krngd2** [comment]
> What we really need especially on the iOS side is a way to add a Done button to the keyboard.

there is a package 
https://pub.dartlang.org/packages/keyboard_actions

**Comment by albertgeorge56** [comment]
Is it included in flutter official docs how to dismiss keyboard when a user tap outside? Also clicking on the back button on android doesn't unfocus the textfield. I have used FocusNode but nothing really worked out for me. I request flutter team to take a look on this issue.

**Comment by zoechi** [comment]
@albertgeorge56 
Please consider asking support questions in one of the other channels listed at http://flutter.io/support .

**Comment by vishaldhudum** [comment]
> @manujbahl The problem with the keyboard doesn't work for everyone, for example losing the focus from a text field and assign it to unused focus node doesn't really work(at least for me):
> 
> ` FocusScope.of(context).requestFocus(new FocusNode());`
> 
> One of the solution that could solve your problem is using the SystemChannel where you act on the text input channel and hide it once the Input Field loses the focus:
> 
> ```dart
>             body: GestureDetector(
>                 onTap: () {
>                   SystemChannels.textInput.invokeMethod('TextInput.hide');
>                 },
> ```
> 
> Hope will solve your problem.

This does dismiss the keyboard and keeps the keyboard if user taps on another text field(which was my issue), but the caret remains at its position in the text field. Anything for this.

**Comment by DeRealMorgan** [comment]
There must be an official way to achieve this simple thing. Has anyone found a more elegant solution by now?

**Comment by atereshkov** [comment]
```
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(AppLocalizations.of(context).calculatorAdvancedStageTitle),
      ),
      body: GestureDetector(
        behavior: HitTestBehavior.opaque,
        onTap: () {
          FocusScope.of(context).requestFocus(new FocusNode());
        },
        child: Padding(
          padding: const EdgeInsets.only(
            left: 14,
            top: 8,
            right: 14,
            bottom: 8,
          ),
          child: Text('Work'),
        ),
      )
    );
  }
```

Works for me.

**Comment by guopeng1994** [comment]
we need more simple ways!!! just list react native ways: Keyboard.dismiss()

**Comment by wuxuan2811** [comment]
import 'package:flutter/services.dart';
SystemChannels.textInput.invokeMethod('TextInput.hide');

**Comment by tallinn1960** [comment]
The best solution seems to me to incorporate the keyboard_actions package into the Form widget or derive a widget from Form that incorporates this. I've seen this in native iOS apps. Isn't it part of the OS? Does Flutter prevent us from using this?  Note that the gesture solution will not dismiss the keyboard on switching tabbed pages.

Those were the days, my friend , we thought they never end .... when focus management and navigation was part of form managers. Got forgotten somehow in website programming where users have been trained to click on fields to give them focus. I cannot count how often I entered a password into everything else but the password field, sending it to everyone interested. Now, on smartphones, you do not only miss an easy way to tab from field to field but to put away your keyboard as well. If that's progress I am getting too old for this. Should I quit learning how to program smartphones? It feels like it.

**Comment by gspencergoog** [comment]
In #31614, I added an `unfocus()` to `FocusNode` and `FocusScopeNode` to allow giving up focus, but it only worked if that node was the primary focus. #31909 changes that so that it will unfocus the entire chain, not just the primary focus. 

Once #31909 lands,  if you call `unfocus()` on a `FocusNode` or `FocusScopeNode`, and their `hasFocus` was returning true, then after calling `unfocus()`, it will return false, and the keyboard should dismiss on text fields.

Once that PR lands, you should be able to simply do "FocusScope.of(context).unfocus()" to dismiss the keyboard/give up focus on a node.

**Comment by marcglasberg** [comment]
@gspencergoog The problem is that, as of now, gaining/losing focus and opening/closing the keyboard are strongly coupled, and they shouldn't be. There is also the related question of letting the widget get focus without letting the keyboard open in the first place (https://github.com/flutter/flutter/issues/16863). This must also be fixed by a change in the `FocusNode`. 

It could have two more methods besides `focus` and `unfocus`. I suggest `focusWithoutKeyboard()` (which would give the TextField the focus without opening the keyboard) and `closeKeyboard()` (which would close the keyboard without removing focus from the TextField). 

Besides, it needs a mutable field called `opensKeyboard`, which is `true` by default. When the `TextField` gains focus by the user tapping it, it would check this field to decide if it should open the keyboard or not.

**Comment by ch-muhammad-adil** [comment]
> @manujbahl The problem with the keyboard doesn't work for everyone, for example losing the focus from a text field and assign it to unused focus node doesn't really work(at least for me):
> 
> ` FocusScope.of(context).requestFocus(new FocusNode());`
> 
> One of the solution that could solve your problem is using the SystemChannel where you act on the text input channel and hide it once the Input Field loses the focus:
> 
> ```dart
>             body: GestureDetector(
>                 onTap: () {
>                   SystemChannels.textInput.invokeMethod('TextInput.hide');
>                 },
> ```
> 
> Hope will solve your problem.

It has helped me, in my project I have been using 

`FocusScope.of(context).requestFocus(FocusNode())` all over the project but at only one screen it is not working and switching focus between different input fields, even selecting two fields at the same time. I am using latest version of the Flutter 1.5+ , 

In my case only one screen is not working by using 

`FocusScope.of(context).requestFocus(FocusNode())`

**Comment by gspencergoog** [comment]
Now that #31909 (be75fb36c5dd0848aa5adfe87c6d54c7174004c4) has landed, you should use `FocusScope.of(context).unfocus()` instead of `FocusScope.of(context).requestFocus(FocusNode())`, since FocusNodes are `ChangeNotifiers`, and should be disposed properly.

(Of course, that's only true for channels that include that change)

**Comment by ch-muhammad-adil** [comment]
Running flutter doctor...
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, v1.5.4-hotfix.2, on Mac OS X 10.14.4 18E226, locale en-PK)
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] iOS toolchain - develop for iOS devices (Xcode 10.1)
[✓] Android Studio (version 3.4)
[✓] Connected device (1 available)

And all I see cursor is getting focused at multiple input fields in Form.

**Comment by ch-muhammad-adil** [comment]
@gspencergoog method is not defined , 
<img width="513" alt="Screenshot 2019-05-10 at 12 47 32 AM" src="https://user-images.githubusercontent.com/8264639/57482111-4562d080-72bd-11e9-9c49-dc93cfec07ca.png">

**Comment by ch-muhammad-adil** [comment]
Checkout how in this small GIF, everything was fine before version **1.5.4-hotfix.2**

![Screen Recording 2019-05-10 at 12 49 50 AM](https://user-images.githubusercontent.com/8264639/57482794-f9189000-72be-11e9-8218-ff44deddd18d.gif)

**Comment by ch-muhammad-adil** [comment]
Here i have uploaded a video of the problem happening with me at latest version of the flutter, It is happening only at full screen forms , 

https://drive.google.com/file/d/1iwXdrRa1rrs1fhe7tgeY7rwyLRLtyVA-/view?usp=sharing

**Comment by gspencergoog** [comment]
Correct, `unfocus` is not available on stable yet.

**Comment by ch-muhammad-adil** [comment]
Here is another video I have created, I have used `unfocus` as well after switching to master channel, but it is also not working , 

https://drive.google.com/file/d/1GteaG0vFa32wxjwJE13ULE86OEhvkr6l/view?usp=sharing

**Comment by ch-muhammad-adil** [comment]
> Correct, `unfocus` is not available on stable yet.

it is not working , at master even. :(

**Comment by ch-muhammad-adil** [comment]
Got resolved, I have been doing a mistake which causing problem to assign focus at multiple TextInputFields , My TextField are actually custom widgets created by me and in case if focusNode is null i was assign it a default FocusNode(), and this have been bringing issues.  

`focusNode: focusNode != null ? focusNode : FocusNode(),`

**Comment by gspencergoog** [comment]
I'm glad you were able to resolve it.  Yes, indeed, constructing a "throw away" `FocusNode` isn't the way to do things anymore (at least on the master channel, soon on the others).

**Comment by C-gotoh** [comment]
@gspencergoog What about

`FocusScope.of(context).detach()`

I have used this to remove the focus from a widget instead of creating a new FocusNode.
Will it be better to use the new ".unfocus()" method when it becomes available in stable?

**Comment by gspencergoog** [comment]
@C-gotoh, In the new implementation, `detach` has moved to a new [`FocusAttachment`](https://master-api.flutter.dev/flutter/widgets/FocusAttachment-class.html) object which maintains the attachment of the node to the focus tree. It will indeed unfocus the node if the node is the primary focus, but will also detach the node from the focus tree, meaning that you will have to call `attach` on it again to get a new `FocusAttachment` object in order to keep using it.

So, if you're trying to detach the node because it's no longer needed, then call `detach`.  If you're trying to just release focus on the node, but intend to keep it around, call `unfocus`.

**Comment by iStiLLinIceAge** [comment]
Try this dude..


import 'package:flutter/material.dart';

void main(){
  runApp(
      MaterialApp(
      home:Test()
    )
  );
}

class Test extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: new Text('Test'),
      ),
      body: new InkWell(
        splashColor: Colors.transparent,
        highlightColor: Colors.transparent,
        onTap: (){
          FocusScope.of(context).requestFocus(new FocusNode());
        },
        child: new Center(
          child: SingleChildScrollView(
            child: Column(
              children: <Widget>[
                Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: new Text('Try this'),
                ),
                Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: new TextFormField(),
                ),
                Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: new TextFormField(),
                ),
              ],
            ),
          )
        ),   
      ),
    );
  }
}

**Comment by iStiLLinIceAge** [comment]
Try this dude...

```
@override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: new Text('Test'),
      ),
      body: new InkWell(
        splashColor: Colors.transparent,
        highlightColor: Colors.transparent,
        onTap: (){
          FocusScope.of(context).requestFocus(new FocusNode());
        },
        child: new Center(
          child: SingleChildScrollView(
            child: Column(
              children: <Widget>[
                Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: new Text('Try this'),
                ),
                Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: new TextFormField(),
                ),
                Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: new TextFormField(),
                ),
              ],
            ),
          )
        ),   
      ),
    );
  }
```

**Comment by chornbe** [comment]
Can you add a native bridge in Flutter / Dart? 'Cause this is trivial in native code. Surprised this didn't bubble up long before now.

**Comment by Xgamefactory** [comment]
why issue has closed ?
after with both workarounds textfield not getging keyboard opened

**Comment by jamesdixon** [comment]
Trying to use to hide the keyboard when tapping away from a form field. To do this, I tried wrapping my entire app in a `GestureDetector` and then calling `FocusScope.of(context).unfocus()`.

```dart
class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: () {
        FocusScope.of(context).unfocus();
      },
      child: MaterialApp(
        title: 'Scout',
        theme: theme,
        home: LoginScreen(),
        routes: routes,
      ),
    );
  }
}
```

Unfortunately, if you tap the body more than once, I get the following error:

```
flutter: The following assertion was thrown while handling a gesture:
flutter: Node has primary focus, but no enclosingScope.
flutter: 'package:flutter/src/widgets/focus_manager.dart':
flutter: Failed assertion: line 549 pos 14: 'scope != null'
```

Is this poor usage of `unfocus()`? Alternatives?

Thanks!

**Comment by gspencergoog** [comment]
What this means, I think, is that you're trying to unfocus the root node in the focus tree.

You can check to see if `FocusScope.of(context)` returns the root node by comparing it against `WidgetsBinding.instance.focusManager.rootScope`.

Perhaps we should make that a no-op instead of an assertion (for the root node).

**Comment by jamesdixon** [comment]
@gspencergoog thanks for the reply! Very much appreciate it.

It seems that when tapping outside of the form field, what's returned by `FocusScope.of(context)` is the root scope, and therefore, `unfocus()` is never called. This is all assuming I understood you correctly 😄 

```dart
class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: () {
        FocusScopeNode currentScope = FocusScope.of(context);
        FocusScopeNode rootScope =
            WidgetsBinding.instance.focusManager.rootScope;

        if (currentScope != rootScope) {
          currentScope.unfocus();
        }
        // FocusScope.of(context).requestFocus(FocusNode());
      },
      child: MaterialApp(
        title: 'Scout',
        theme: theme,
        home: LoginScreen(),
        routes: routes,
      ),
    );
  }
}
```
Is there a method to get the previously focused `FocusNode`?

**Comment by deepakpawade** [comment]
@gspencergoog thank .. the code solved my problem...

**Comment by jamesdixon** [comment]
This works as expected without throwing any exceptions:

```dart
    return GestureDetector(
      onTap: () {
        FocusScopeNode currentFocus = FocusScope.of(context);
        if (!currentFocus.hasPrimaryFocus) {
          currentFocus.unfocus();
        }
      },
      child: ...
```

**Comment by jifalops** [comment]
I've tried several things here such as 

```dart
    FocusScopeNode currentFocus = FocusScope.of(context);
    if (currentFocus != WidgetsBinding.instance.focusManager.rootScope &&
        !currentFocus.hasPrimaryFocus) {
      currentFocus.unfocus();
    }
```
and
```dart
WidgetsBinding.instance.focusManager.primaryFocus?.unfocus();
```

The only thing that has worked for me is to use `FocusScope.of(context).requestFocus(FocusNode())`, but it throws a concurrent modification exception if you call that in response to a text field being tapped.

So the final working solution was:

```dart
Future.microtask(() => FocusScope.of(context).requestFocus(FocusNode()));
```

**Comment by jamesdixon** [comment]
@jifalops have you tried without `currentFocus != WidgetsBinding.instance.focusManager.rootScope`?

I just wrote an [article on this topic](https://flutter360.dev/dismiss-keyboard-form-lose-focus/) that contains the full code. I'm curious if there's something different about your approach as this works for me.

**Comment by jifalops** [comment]
Yes I tried it without checking the `rootScope`, but from what I gather, calling unfocus() on the root scope causes an exception. My use case involves hiding the keyboard when the text field _gains_ focus, which is a bit more stringent than most use cases.

**Comment by jifalops** [comment]
@jamesdixon That is a clever way to unfocus any input in the app when tapping outside of them.

**Comment by jamesdixon** [comment]
@jifalops thanks! out of curiosity, what's your use case for hiding the keyboard when the form field gains focus?

**Comment by jifalops** [comment]
To show the date/time picker dialogs instead of initially going into edit mode. https://pub.dev/packages/datetime_picker_formfield

**Comment by jamesdixon** [comment]
Nice. Cool package! Going to give it a shot 👍

**Comment by JamesMcIntosh** [comment]
@jifalops @jamesdixon If `unfocus()` is not working then pay VERY close attention to which BuildContext you have passed in to the `FocusScope.of(context)`.
In particular I made the mistake of using the BuildContext passed into a build method rather than the context supplied from a Builder function, my particular case the function was building up a MaterialPageRoute so the context was well out of expected scope by the time the builder was called i.e.
```
MaterialPageRoute<void> buildRoute(BuildContext context, Widget child) {
  return MaterialPageRoute<void>(
    builder: (BuildContext builderContext) => GestureDetector(
      onTap: () {
        FocusScopeNode currentFocus = FocusScope.of(context); // SHOULD BE builderContext
        if (!currentFocus.hasPrimaryFocus) {
          currentFocus.unfocus();
        }
      },
      child: child,
    ),
  );
}
```

**Comment by awkaw** [comment]
`
import 'package:flutter/services.dart';

SystemChannels.textInput.invokeMethod('TextInput.hide');
`

**Comment by BahijMik** [comment]
Is there any way to actually manipulate the keyboard hide button on android? just like in android there is onConfigurationChanged, i want to decrease/increase image size when the keyboard is shown/hidden.

**Comment by malviyaritesh** [comment]
As mentioned in https://github.com/flutter/flutter/issues/7247#issuecomment-488307127
Now, simply you can `unfocus` also:
```dart
FocusScope.of(context).unfocus();
```

**Comment by MichelRahme** [comment]
If i have a `TextField` implemented in a Stack with `GoogleMaps` widget, 
`FocusScope.of(context).unfocus();` will not work. Any Solution ?

**Comment by TKH219** [comment]
> If i have a `TextField` implemented in a Stack with `GoogleMaps` widget,
> `FocusScope.of(context).unfocus();` will not work. Any Solution ?

WidgetsBinding.instance.focusManager.primaryFocus?.unfocus(); 
You should try it.

**Comment by Ladriel** [comment]
There is a few points to mention and put into light here 

--------------
Nobody ever mentioned that the solution proned by everyone to reset the focus basically conflicts with text editing options (the 'copy/paste/select all' menus).

For exemple :
Hold a tap on a text to make the contextual text editing options appear on your device and try to use the 'select all' option, your focus is now reseted because of the fix and the text is unselected.

For this particular concerns the preferred method for me has been :
SystemChannels.textInput.invokeMethod('TextInput.hide');

-----------------
We dropped it since first iterations but i already been experiencing with reactions to long press by wrapping the whole app with a Listener with onPointerDown and onPointerUp event listeners. You can evaluate the timing between down and up events to recognize and handle long press too.

A way to know which particular Widget is the target of focus on tap would come in really handy too.
Something like :
'FocusScope.of(context).componentType'  which would answer 'TextFormField'.
I couldn't find a way to get this informations (from my experimentations at least, correct me if i'm wrong).

------------

Also the missing Done button on iOs numeric keyboards in Flutter is kind of a problem, your one and only option to dismiss the keyboard is then to use the solutions detailed in this thread. Which does seem kind of hacky-wacky.

------------

Thought i'd leave these points here if anybody would ever need them.

**Comment by gspencergoog** [comment]
@Ladriel, you can indeed get the runtime type of the widget that has the current focus:

```dart
Type focusedType = Focus.of(context).context.widget.runtimeType;
```

But this is a question that would be better asked on [Stack Overflow](https://stackoverflow.com/questions/tagged/flutter).

As for your other comments, please file separate bugs for separate issues.

I'm going to close this bug for comments, since the issue database isn't really the best place for conversation or questions: the bug has been closed for a long time.
